api/database_engine.go

```diff
@@ -67,7 +67,147 @@ func (e *EverestServer) UpdateDatabaseEngine(ctx echo.Context, namespace, name s
+		engines, err := e.kubeClient.ListDatabaseEngines(ctx, namespace)
+		if err != nil {
+			return err
+		}
+		for _, engine := range engines.Items {
+			check, err := e.checkDatabases(ctx, &engine)
+			if err != nil {
+				return err
+			}
+			for _, c := range check {
+				*result.PendingActions = append(*result.PendingActions, UpgradeTask{
+					Name:        c.Name,
+					PendingTask: pointer.To(UpgradeTaskPendingTask(pointer.Get(c.PendingTask))),
+					Message:     c.Message,
+				})
+			}
+		}
+	}
+	return c.JSON(http.StatusOK, result)
+}
+
+// ApproveUpgradePlan starts the upgrade of operators in the provided namespace.
+func (e *EverestServer) ApproveUpgradePlan(c echo.Context, namespace string) (err error) {
+	ctx := c.Request().Context()
+
+	up, err := e.getUpgradePlan(ctx, namespace)
+	if err != nil {
+		return err
+	}
+
+	// lock all engines that will be upgraded.
+	if err := e.setLockDBEnginesForUpgrade(ctx, namespace, up, true); err != nil {
+		return errors.Join(err, errors.New("failed to lock engines"))
+	}
+	// unlock if we return with error.
+	defer func() {
+		if err != nil {
+			if lockErr := e.setLockDBEnginesForUpgrade(ctx, namespace, up, false); lockErr != nil {
+				err = errors.Join(err, errors.New("failed to unlock engines"))
+			}
+		}
+	}()
+
+	// Check if we're ready to upgrade?
+	if slices.ContainsFunc(pointer.Get(up.PendingActions), func(task UpgradeTask) bool {
+		return pointer.Get(task.PendingTask) != Ready
+	}) {
+		return c.JSON(http.StatusPreconditionFailed, Error{
+			Message: pointer.ToString("One or more database clusters are not ready for upgrade"),
+		})
+	}
+
+	// start upgrade process.
+	if err := e.startOperatorUpgradeWithRetry(ctx, "", namespace, ""); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *EverestServer) getUpgradePlan(
+	ctx context.Context,
+	namespace string,
+) (*UpgradePlan, error) {
+	engines, err := e.kubeClient.ListDatabaseEngines(ctx, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	result := &UpgradePlan{
+		Upgrades:       pointer.To([]Upgrade{}),
+		PendingActions: pointer.To([]UpgradeTask{}),
+	}
+
+	for _, engine := range engines.Items {
+		nextVersion := engine.Status.GetNextUpgradeVersion()
+		if nextVersion == "" {
+			continue
+		}
+
+		upgrade := &Upgrade{
+			CurrentVersion: pointer.To(engine.Status.OperatorVersion),
+			Name:           pointer.To(engine.GetName()),
+			TargetVersion:  pointer.To(nextVersion),
+		}
+		*result.Upgrades = append(*result.Upgrades, *upgrade)
+		pf, err := e.getOperatorUpgradePreflight(ctx, nextVersion, &engine)
+		if err != nil {
+			return nil, err
+		}
+		for _, db := range pointer.Get(pointer.Get(pf).Databases) {
+			*result.PendingActions = append(*result.PendingActions, db.toUpgradeTask())
+		}
+	}
+	return result, nil
+}
+
+// ToOperatorUpgradeDatabaseItem converts OperatorUpgradePreflightForDatabase to OperatorUpgradeDatabaseItem.
+// TODO: Remove this function when the deprecated API is removed.
```

Comment:
Can this function be removed? It's part of the new solution...

Reply:
I think I forgot to remove the TODO there, but this was supposed to be a helper. I can still remove it and in-line it on the caller. Although, what's wrong in keeping this?

Reply:
Nothing wrong, I just thought it was an incorrect comment because it's actually being used in the new solution, if you plan to inline it that's fine.